### PR TITLE
Implemented lazy logging by replacing isDebugEnabled() with lambda invocations instead of checking explicitly 

### PR DESCRIPTION
**Pull request by satti-hari-krishna-reddy**
### Description
This pull request aims to enhance the logging mechanism by replacing multiple instances of `isDebugEnabled()` with lambda-based logging. This change improves the code's efficiency and readability by avoiding unnecessary string concatenation and evaluating expensive operations when debug logging is disabled.

### Related Issues
Resolves #[8646]

### CODE REVIEW

File: client/rest/src/main/java/org/opensearch/client/RequestLogger.java [hunk 1]
```diff
@@ -67,11 +68,10 @@ private RequestLogger() {}
      * Logs a request that yielded a response
      */
     static void logResponse(Log logger, HttpUriRequest request, HttpHost host, ClassicHttpResponse httpResponse) {
-        if (logger.isDebugEnabled()) {
-            logger.debug(
-                "request [" + request.getMethod() + " " + host + getUri(request) + "] returned [" + new StatusLine(httpResponse) + "]"
-            );
-        }
+        logger.debug("request [{} {} {}] returned [{}]", () -> request.getMethod(), () -> host,
```
Comment (dbwiddis) [inline client/rest/src/main/java/org/opensearch/client/RequestLogger.java:71]:
```suggestion
        logger.debug("request [{} {}{}] returned [{}]", () -> request.getMethod(), () -> host,
```
In original there's no space between host and URI

File: client/rest/src/main/java/org/opensearch/client/RequestLogger.java [hunk 2]
```diff
@@ -101,9 +101,9 @@ static void logResponse(Log logger, HttpUriRequest request, HttpHost host, Class
      * Logs a request that failed
      */
     static void logFailedRequest(Log logger, HttpUriRequest request, Node node, Exception e) {
-        if (logger.isDebugEnabled()) {
-            logger.debug("request [" + request.getMethod() + " " + node.getHost() + getUri(request) + "] failed", e);
-        }
+        logger.debug("request [{} {} {}] failed", () -> request.getMethod(), () -> node.getHost(),
```
Comment (dbwiddis) [inline client/rest/src/main/java/org/opensearch/client/RequestLogger.java:104]:
```suggestion
        logger.debug("request [{} {}{}] failed", () -> request.getMethod(), () -> node.getHost(),
```

File: server/src/main/java/org/opensearch/snapshots/SnapshotShardsService.java [hunk 3]
```diff
@@ -292,16 +292,14 @@ private void startNewShards(SnapshotsInProgress.Entry entry, Map<ShardId, IndexS
...
                                 assert newGeneration.equals(snapshotStatus.generation());
-                                if (logger.isDebugEnabled()) {
-                                    final IndexShardSnapshotStatus.Copy lastSnapshotStatus = snapshotStatus.asCopy();
-                                    logger.debug(
-                                        "snapshot [{}] completed to [{}] with [{}] at generation [{}]",
-                                        snapshot,
-                                        snapshot.getRepository(),
-                                        lastSnapshotStatus,
-                                        snapshotStatus.generation()
+                                final IndexShardSnapshotStatus.Copy lastSnapshotStatus = snapshotStatus.asCopy();
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/snapshots/SnapshotShardsService.java:295]: This change moves this new object creation outside of the isDebug check. Is there a way to move it down into the lambda where it's used?

File: server/src/main/java/org/opensearch/snapshots/SnapshotShardsService.java [hunk 4]
```diff
@@ -292,16 +292,14 @@ private void startNewShards(SnapshotsInProgress.Entry entry, Map<ShardId, IndexS
...
+                                logger.debug(
+                                    "snapshot [{}] completed to [{}] with [{}] at generation [{}]",
+                                     () -> snapshot,
+                                     () -> snapshot.getRepository(),
+                                     () -> lastSnapshotStatus,
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/snapshots/SnapshotShardsService.java:300]:
```suggestion
                                     () -> snapshotStatus.asCopy(),
```
(Not sure this is correct but is what I meant by the previous comment.)

**Review by dbwiddis — COMMENTED** [review]
A few minor nits.

**Comment by satti-hari-krishna-reddy** [comment]
hi @dbwiddis I appreciate your feedback and made the suggested change. Could you please take a quick look to let me know if it's on the right track?

**Comment by sandeshkr419** [comment]
@satti-hari-krishna-reddy Thanks for starting working on this. 

Few key points:
1. Try to ensure that the modified code produces the exact same log output - no changes in uppercase/lowercase, whitespaces, indentations, formatting, etc.
2. Look into CI failures as well - your gradle failure most likely should be due to stale changes - try rebasing against main branch.

**Comment by satti-hari-krishna-reddy** [comment]
@sandeshkr419   I went through the code again and tried to match the log output as you mentioned. I also looked into the CI failures, and I rebased against the main branch. However, the CI is still failing. I think I might have messed up somewhere. Can you please help me figure out what the issue might be? Thanks for your assistance.

**Comment by andrross** [comment]
@satti-hari-krishna-reddy Java compilation is failing. See the [Jenkins output](https://build.ci.opensearch.org/job/gradle-check/23559/) linked above. You can recreate the failure locally with `./gradlew :client:rest:compileJava`.

File: client/rest/src/main/java/org/opensearch/client/RequestLogger.java [hunk 5]
```diff
@@ -67,11 +68,10 @@ private RequestLogger() {}
...
     static void logResponse(Log logger, HttpUriRequest request, HttpHost host, ClassicHttpResponse httpResponse) {
-        if (logger.isDebugEnabled()) {
-            logger.debug(
-                "request [" + request.getMethod() + " " + host + getUri(request) + "] returned [" + new StatusLine(httpResponse) + "]"
-            );
-        }
+        logger.debug("request [{} {}{}] returned [{}]", () -> request.getMethod(), () -> host,
+        () -> getUri(request),
+        () -> new StatusLine(httpResponse));
+
```
Comment (Bukhtawar) [inline client/rest/src/main/java/org/opensearch/client/RequestLogger.java:74]: why not use 
```
logger.debug(() -> new ParameterisedMessage("Log blah {}"), blah); 
```

Comment (sandeshkr419) [reply client/rest/src/main/java/org/opensearch/client/RequestLogger.java:74]: Hi @Bukhtawar, Thanks for sharing inputs.

The lambda expression itself is responsible for creating a ParameterizedMessage instance and formatting the log message. So, in the statement `logger.debug(() -> new ParameterizedMessage("Log blah {}", blah));` you suggested, `ParameterizedMessage` constructor and message formatting will be executed when this statement is executed, regardless of the log level.

However, in `logger.debug("Log blah {}", () -> blah)`, the message template "Log blah {}" is provided as a constant string and the lambda expression `() -> blah` will only be executed if the debug log level is enabled. This allows for lazy computation, as `blah` is evaluated only when necessary. 

The lazy computation is what is desired, hence, I'd recommend the way I originally recommended. If there is some information which you think I have not accounted for, please feel free to share.

**Comment by sandeshkr419** [comment]
Hi @satti-hari-krishna-reddy, were you able to understand failures by reproducing it i local as @andrross suggested above? Please let me know if you are still blocked?

**Comment by sandeshkr419** [comment]
Hi @satti-hari-krishna-reddy, just want check if you are still working on this or if you need any help getting unblocked? Feel free to ping here or on our slack workspace, we'll be happy to assist in overcoming any blockers. 

Thanks!

**Comment by kotwanikunal** [comment]
The PR has been stalled for quite a while. Please reopen if you are still working on it.

**Pull request closed without merging**